pkg/http/delayed_writer.go

```diff
@@ -0,0 +1,67 @@
+package http
+
+import (
+	"net/http"
+)
+
+var (
+	_ http.Flusher        = (*DelayedWriter)(nil)
+	_ http.ResponseWriter = (*DelayedWriter)(nil)
+)
+
+// DelayedWriter is an implementation of http.ResponseWriter. It delays writing to the output until RealFlush is called.
+type DelayedWriter struct {
+	ResponseCode int
+
+	writer         http.ResponseWriter
+	headerRecorded bool
+
+	buffer        []byte
+	bytesBuffered int
+}
+
+// NewDelayedWriter creates an http.ResponseWriter that delays writing.
+func NewDelayedWriter(w http.ResponseWriter) *DelayedWriter {
+	return &DelayedWriter{
+		writer: w,
+	}
+}
+
+// Flush does nothing.
+func (dw *DelayedWriter) Flush() {}
+
+// RealFlush flushes the buffer (both header and body) to the client.
+func (dw *DelayedWriter) RealFlush() {
+	dw.writer.WriteHeader(dw.ResponseCode)
+	_, err := dw.writer.Write(dw.buffer[:dw.bytesBuffered])
+	if err != nil {
+		panic(err)
```

Comment:
we probably don't want panic here, an error message?

Reply:
That makes sense! I'll change it to an error message